File: Linear Control Course/Chapter4/Lesson5/Chapter4_Lesson5.m
% Plant and sensor dynamics, e.g. linearized robotic joint
s = tf('s');
G = 10 / (s * (s + 2));   % actuator dynamics
H = 0.5;                  % sensor gain

% Closed-loop TF via Mason (single forward path, single loop)
T_mason = G / (1 + G*H);

% Closed-loop TF via built-in feedback function (negative feedback)
T_matlab = feedback(G, H, -1);

% Verify equality
T_simplified = minreal(T_mason - T_matlab)

File: Tutorials/Codes/Linear_Control_Course/Chapter4/Lesson5/Chapter4_Lesson5.py
import sympy as sp

# Symbolic Laplace variable
s = sp.symbols('s')

# Example plant and sensor dynamics (e.g. simple robotic actuator plus encoder)
# G(s) = 10 / (s (s + 2)),   H(s) = 0.5
G = 10 / (s * (s + 2))
H = sp.Rational(1, 2)

# Mason's formula for single forward path and single loop:
# P1 = G, L1 = -G*H, Delta = 1 - L1 = 1 + G*H, Delta1 = 1
T_mason = sp.simplify(G / (1 + G*H))
print("Closed-loop TF from Mason:", T_mason)

# Numerical check using python-control (commonly used in robotics)
import control as ctrl

# Plant and sensor transfer functions in python-control
G_tf = ctrl.TransferFunction([10], [1, 2, 0])   # 10 / (s^2 + 2 s)
H_tf = ctrl.TransferFunction([0.5], [1])        # 0.5

# Negative feedback (unity reference, sensor H)
T_tf = ctrl.feedback(G_tf, H_tf, sign=-1)
print("Closed-loop TF from python-control:", T_tf)

# The rational form of T_tf should match the symbolic T_mason after simplification.
